doc/ni_by_example_2.md

```diff
@@ -0,0 +1,395 @@
+
+
+To start this tutorial, At this point, it's important to have a general idea of how `ni` works
+
+##`ni` is a quine
+
+A quine is a program that prints its source code. 
+
+
+##`ni` Philosophy and Style
+
+####`ni` demands expertise
+
+####Conciseness matters; readability to the uninitiated does not.
+`ni` spells should be beautiful
+
+####Outsource hard jobs to more appropriate tools.
+The most obvious 
+
+Some jobs that are difficult for `ni`:
+* Sorting
+* Matrix Multiplication
+* SQL Joins
+
+
+
+
+##Perl for `ni` fundamentals
+
+
+
+##Connecting `bash` and `ni`
+
+* `e[<script>]`: Evaluate script
+  * evaluate `<script>` in bash, and stream out the results one line at a time.
+
+##Input Operations
+* `D:<field1>,:<field2>...`: JSON Destructure
+  * `ni` implements a very fast JSON parser that is great at pulling out string and numeral fields.
+  * As of 2016-12-24, the JSON destructurer does not support list-based fields in JSON.
+
+####`ni` and directories
+
+Enrichment sections explore relevant but non-critical `ni` functions. Their contents are not critical to productive `ni` development, and they can be skipped or skimmed.
+
+Start by making some data (Note that you have to be in `bash` for the echo statements to work. [Here](http://stackoverflow.com/questions/8467424/echo-newline-in-bash-prints-literal-n) is a very interesting post about `echo`'s unintuitive behavior):
+
+```
+$ rm -rf dir_test && mkdir dir_test
+$ echo -e "hello\n" > dir_test/hi
+$ echo -e "you\n" > dir_test/there
+```
+
+Let's start with `$ ni test`
+
+```
+dir_test/hi
+dir_test/there
+(END)
+```
+
+`ni` has converted the folder into a stream of the names of files (and directories) inside it. You can thus access the files inside a directory using `\<`.
+
+`$ ni dir_test \<`
+
+yields:
+
+```
+hello
+you
+(END)
+```
+
+`ni` also works with the bash expansion operator `*`.
+
+For example, `ni dir_test/*` yields:
+
+```
+hello
+you
+(END)
+```
+
+`ni` is able to go to the files directly becasue it is applies bash expansion first; bash expansion generates the file paths, which `ni` is then able to interpret.
+
+```
+$ ni --explain dir_test/*
+["cat","dir_test/hi"]
+["cat","dir_test/there"]
+```
+
+
+  
+##Understanding the `ni` monitor
+More details [here](monitor.md). Overall:
+
+* Negative numbers = non-rate-determining step
+* Positive numbers = rate-determining step
+* Large Positive numbers = slow step
+
+
+##SSH and Containers
+
+* `C<container_name>[...]`: execute `[...]` in `<container_name>`
+  * Running in containers requires that Docker be installed on your machine.
+  * Running containers can be especially useful to take advantage of better OS-dependent utilities.
+  * For example, Mac OS X's `sort` is painfully slow compared to Ubuntu's. If you are developing on a Mac, there will be a noticeable performance difference increase by replacing:
```

Comment:
It's possible that you can also address this by using `brew` or equivalent to install gnu coreutils.